Drop FutureNCCL in favor of vanilla CUDAFuture

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #48840 Remove DataPtr extractor from CUDAFuture
* **#49014 Drop FutureNCCL in favor of vanilla CUDAFuture**

We extracted a generic and reusable CUDAFuture class from FutureNCCL, but we had left FutureNCCL around, as a subclass of CUDAFuture, in order to deal with some peculiarity of ProcessGroupNCCL, namely that the future would be completed right away when constructed and that its CUDA events would be _shared_ with the ones of the WorkNCCL. This required some "hacks" in CUDAFuture itself (protected members, fields wrapped in shared_ptrs, ...).

My understanding is that creating CUDA events is a rather cheap operation. That would mean that we could afford to record _twice_ the events after each NCCL call, once for the WorkNCCL and once for the future. By doing so, we can use the CUDAFuture class directly and revert all its hacks.

Differential Revision: [D25355272](https://our.internmc.facebook.com/intern/diff/D25355272/)

lw committed:
Drop FutureNCCL in favor of vanilla CUDAFuture

We extracted a generic and reusable CUDAFuture class from FutureNCCL, but we had left FutureNCCL around, as a subclass of CUDAFuture, in order to deal with some peculiarity of ProcessGroupNCCL, namely that the future would be completed right away when constructed and that its CUDA events would be _shared_ with the ones of the WorkNCCL. This required some "hacks" in CUDAFuture itself (protected members, fields wrapped in shared_ptrs, ...).

My understanding is that creating CUDA events is a rather cheap operation. That would mean that we could afford to record _twice_ the events after each NCCL call, once for the WorkNCCL and once for the future. By doing so, we can use the CUDAFuture class directly and revert all its hacks.

Differential Revision: [D25355272](https://our.internmc.facebook.com/intern/diff/D25355272/)

[ghstack-poisoned]

mrshenli reviewed (changes requested):
test failures look relevant on this one

mrshenli commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -190,62 +194,6 @@ class ProcessGroupNCCL : public ProcessGroup {
-  // enables synchronizing the appropriate streams and avoids stalling PyTorch's
-  // default stream while running the callback. In case of multiple then
-  // callbacks, each will be executed on its own fresh stream.
-  struct FutureNCCL : at::cuda::CUDAFuture {
```
nice!

mrshenli replied:
`FutureNCCL` is also mentioned in our docs. Let update those too:

https://github.com/pytorch/pytorch/blob/dee82ef3ea1ed2cae4e891f071adff03cb4a6e0d/torch/csrc/distributed/c10d/init.cpp#L1140-L1144

mrshenli reviewed (commented):
Rest of the change LGTM!

lw committed:
Update on "Drop FutureNCCL in favor of vanilla CUDAFuture"


We extracted a generic and reusable CUDAFuture class from FutureNCCL, but we had left FutureNCCL around, as a subclass of CUDAFuture, in order to deal with some peculiarity of ProcessGroupNCCL, namely that the future would be completed right away when constructed and that its CUDA events would be _shared_ with the ones of the WorkNCCL. This required some "hacks" in CUDAFuture itself (protected members, fields wrapped in shared_ptrs, ...).

My understanding is that creating CUDA events is a rather cheap operation. That would mean that we could afford to record _twice_ the events after each NCCL call, once for the WorkNCCL and once for the future. By doing so, we can use the CUDAFuture class directly and revert all its hacks.

Differential Revision: [D25355272](https://our.internmc.facebook.com/intern/diff/D25355272/)

[ghstack-poisoned]

lw committed:
Update on "Drop FutureNCCL in favor of vanilla CUDAFuture"


We extracted a generic and reusable CUDAFuture class from FutureNCCL, but we had left FutureNCCL around, as a subclass of CUDAFuture, in order to deal with some peculiarity of ProcessGroupNCCL, namely that the future would be completed right away when constructed and that its CUDA events would be _shared_ with the ones of the WorkNCCL. This required some "hacks" in CUDAFuture itself (protected members, fields wrapped in shared_ptrs, ...).

My understanding is that creating CUDA events is a rather cheap operation. That would mean that we could afford to record _twice_ the events after each NCCL call, once for the WorkNCCL and once for the future. By doing so, we can use the CUDAFuture class directly and revert all its hacks.

Differential Revision: [D25355272](https://our.internmc.facebook.com/intern/diff/D25355272/)

[ghstack-poisoned]

lw commented:
I think I have fixed the build failures

mrshenli reviewed (approved):
LGTM!
